src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs

```diff
@@ -29,13 +30,16 @@ internal abstract partial class AbstractPullDiagnosticHandler<TDiagnosticsParams
     /// the version stamp but not the content (for example, forking LSP text).
     /// </summary>
     private sealed class DiagnosticsPullCache(IGlobalOptionService globalOptions, string uniqueKey)
-        : VersionedPullCache<(int globalStateVersion, VersionStamp? dependentVersion), (int globalStateVersion, Checksum dependentChecksum), DiagnosticsRequestState, ImmutableArray<DiagnosticData>>(uniqueKey)
+        : VersionedPullCache<(int globalStateVersion, VersionStamp? dependentVersion, Checksum? executionVersion), (int globalStateVersion, Checksum dependentChecksum), DiagnosticsRequestState, ImmutableArray<DiagnosticData>>(uniqueKey)
     {
         private readonly IGlobalOptionService _globalOptions = globalOptions;
 
-        public override async Task<(int globalStateVersion, VersionStamp? dependentVersion)> ComputeCheapVersionAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
+        public override async Task<(int globalStateVersion, VersionStamp? dependentVersion, Checksum? executionVersion)> ComputeCheapVersionAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
         {
-            return (state.GlobalStateVersion, await state.Project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false));
+            Checksum? executionVersion = state.Project.Solution.CompilationState.SourceGeneratorExecutionVersionMap.Map.TryGetValue(state.Project.Id, out var sourceGeneratorExecutionVersion) ?
```

Comment:
This was the problem - the cheap version is not incrementing when the sg execution version changed (the expensive one did and does, that was added in https://github.com/dotnet/roslyn/pull/77271/).

The rest of the refresh support for sg files remains as-is

Reply:
1 seems possible.  Though I'm not actually sure how it would work.  Wouls the sg execution map also map to a Version that changed whenever the value charged?

That actually seems somewhat reasonable...

I do like 2 though. It just removes complexity here.  I'm not concerned about checksum cost. We're constantly computing (and caching those values), so I don't think this fast path is generally that helpful. 

So let's do 2.